pkg/node/address.go

```diff
@@ -408,6 +410,8 @@ func RestoreHostIPs(ipv6 bool, fromK8s, fromFS net.IP, cidr *cidr.CIDR) {
 	case err == nil:
 		setter(ip)
 	}
+
+	return ip
```

Comment:
I assume the case you're talking about is

```
	switch {
	case fromK8s != nil && fromFS != nil:
		if fromK8s.Equal(fromFS) {
			ip = fromK8s
		} else {
			ip = fromFS
			err = errMismatch
		}
```

If the K8s and FS IP are the same, then yes we will throw away the IP, by setting it to `nil` which will trigger code later on in the daemon to allocate a new IP. If they are NOT the same, then we will use the FS IP because we prefer local state over the K8s source. In the case that the FS IP is actually invalid (out of CIDR range), then yes we don't try to use the K8s IP. That is the missing case.

However, you mention that it shouldn't be this nuanced, but I'm not clear why it shouldn't. Shouldn't we try our best to restore the IP and ensure that whichever source it's from, that it is a valid IP given the node setup?

Reply:
> However, you mention that it shouldn't be this nuanced, but I'm not clear why it shouldn't. Shouldn't we try our best to restore the IP and ensure that whichever source it's from, that it is a valid IP given the node setup?

Yes, we should. In principle I think that this is the right goal and I think that the code achieves it today for the most part.

>  In the case that the FS IP is actually invalid (out of CIDR range), then yes we don't try to use the K8s IP. That is the missing case.

I think this is my main concern. I found it a bit tricky to understand the implications of this case. Above I came to the conclusion that this shouldn't be a big deal because it would only occur if you changed the CIDR, but the fact that this case exists makes me feel like the code was written with the "good cases" in mind and the "bad cases" (eg changing CIDR) were ignored. Ideally the path would encompass all possible ways this code could be called.

More concretely, in that case, we return `nil`, then we go into `removeOldRouterState()` and iff there is >1 address assigned to the cilium_host interface, _then_ we will go and remove all IPs from the cilium_host device and leave the overall system in a weird state. The number of conditions that need to occur in order to trigger this bad state is large and I can imagine anyone having a lot of trouble trying to debug how things got to that point.

The other background thought I had is just that this code is trying to be ambivalent about which address is being used - the caller all the way up in the daemon passes exactly one address, IPv4 or IPv6. So in theory this code can just focus on the addresses that are provided to it. And yet, in each of the `removeOldRouterState()` and `RestoreHostIPs()` the code is still trying to handle both the IPv4 and IPv6 cases. So we don't gain the benefits of abstracting away the IP family (ie that the reader doesn't have to worry about the other case), but instead the reader must understand multiple layers of functions all through this area in order to understand the way that IPv4 addresses or IPv6 addresses are handled during restore. This is for instance why I was surprised by the actual result of [this discussion thread](https://github.com/cilium/cilium/pull/17762#discussion_r741478517) where I came to the conclusion that the code is correct, but implicitly because of the interaction between the way that IP addresses are listed from the kernel and the way that this function is called with one particular address family at a time.

Similarly, from this thread when I raise that 'it seems like the `errDoesNotBelong` case seems like the case that `removeOldRouterState()` is trying to handle, I worry that we're duplicating the reaction logic across multiple different functions in ways that are subtly different. If the conditional statements don't 100% line up then it's not obvious whether this is intended because the code was thought through deeply, or because these two pieces of logic were supposed to happen in exactly the same conditions (which implies that they should be run from the same function in the same place).

I guess, as a summary the abstractions here seem a bit leaky. It's probably not going to cause problems for 90% of cases but I wish it were more "obviously correct" rather than needing the reviewer/reader to study the various interactions between 4 or 5 functions to understand whether this whole restore logic works together.

Reply:
We discussed these concerns out-of-band and agreed that the potential bug raised above can be addressed in a new commit (included in the latest version). With everything in place we can consider refactoring to improve the readability of this code, but it's not as crucial and not really appropriate for this bugfix PR.

Reply:
Synced offline about resolving the missing logic, which should be addressed in the latest push. I created https://github.com/cilium/cilium/issues/17771 as a followup.